Fix SVG paint server fallback handling for a non-existent URL

Ahmad-S792 opened the pull request:
#### d8fe46f6db4623f4b1cfff8d264d36cfded2e6d1
<pre>
Fix SVG paint server fallback handling for a non-existent URL

<a href="https://bugs.webkit.org/show_bug.cgi?id=287374">https://bugs.webkit.org/show_bug.cgi?id=287374</a>
<a href="https://rdar.apple.com/144493507">rdar://144493507</a>

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/08f435e092b8e4d8ee158c532489aaab08f8d0c4">https://chromium.googlesource.com/chromium/src.git/+/08f435e092b8e4d8ee158c532489aaab08f8d0c4</a> ,
<a href="https://chromium.googlesource.com/chromium/src.git/+/8a0b30e1432898d243c79bab84b373243edb97f2">https://chromium.googlesource.com/chromium/src.git/+/8a0b30e1432898d243c79bab84b373243edb97f2</a> ,
<a href="https://chromium.googlesource.com/chromium/src.git/+/9ba199165cab706b90519a681f67ecce92241d5d">https://chromium.googlesource.com/chromium/src.git/+/9ba199165cab706b90519a681f67ecce92241d5d</a>

This patch aligns and simplify SVG Paint Server handling as per web specification [1]:

&quot;If a paint server reference in a &lt;paint&gt; is invalid, and no fall-back value is given,
no paint is rendered for that layer.&quot;

&amp;

&quot;The value none is a fallback value to use if the paint server reference in the layer
is invalid (due to pointing to an element that does not exist or which is not a
valid paint server).&quot;

[1] <a href="https://www.w3.org/TR/SVG2/painting.html#SpecifyingPaint">https://www.w3.org/TR/SVG2/painting.html#SpecifyingPaint</a>

This also syncs WPT from upstream for &quot;fill-fallback-invalid-uri&quot; test:

Upstream commit: <a href="https://github.com/web-platform-tests/wpt/commit/5e030604129f588749fba9d39241edcfe9df1d8c">https://github.com/web-platform-tests/wpt/commit/5e030604129f588749fba9d39241edcfe9df1d8c</a>

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResource.cpp:
(WebCore::requestPaintingResource):
(WebCore::inheritColorFromParentStyleIfNeeded): Deleted.
* LayoutTests/platform/glib/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/glib/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/platform/glib/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/platform/gtk/svg/custom/invalid-fill-expected.png: Removed.
* LayoutTests/platform/gtk/svg/custom/invalid-uri-stroke-expected.png: Removed.
* LayoutTests/platform/gtk/svg/hixie/error/003-expected.png: Removed.
* LayoutTests/platform/ios/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/ios/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/platform/ios/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-fill-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-uri-stroke-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/hixie/error/003-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/invalid-fill-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/invalid-uri-stroke-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/hixie/error/003-expected.png: Removed.
* LayoutTests/platform/mac/svg/custom/invalid-fill-expected.png: Removed.
* LayoutTests/platform/mac/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/mac/svg/custom/invalid-uri-stroke-expected.png: Removed.
* LayoutTests/platform/mac/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/platform/mac/svg/hixie/error/003-expected.png: Removed.
* LayoutTests/platform/mac/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/platform/win/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/win/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/svg/custom/invalid-fill.svg: Removed.
* LayoutTests/svg/custom/invalid-uri-stroke.svg: Removed.
* LayoutTests/svg/custom/recursive-pattern.svg:
* LayoutTests/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/svg/hixie/error/003.xml: Removed.
* LayoutTests/platform/gtk/svg/custom/external-paintserver-reference-expected.png: Removed.
* LayoutTests/svg/custom/external-paintserver-reference-expected.txt: Rebaselined
* LayoutTests/platform/glib/svg/custom/recursive-pattern-expected.png: Removed.
* LayoutTests/platform/glib/svg/custom/recursive-pattern-expected.txt:
* LayoutTests/platform/gtk/svg/custom/recursive-pattern-expected.png: Removed.
* LayoutTests/platform/ios/svg/custom/recursive-pattern-expected.txt:
* LayoutTests/platform/mac/svg/custom/recursive-pattern-expected.png: Removed.
* LayoutTests/platform/mac/svg/custom/recursive-pattern-expected.txt:
* LayoutTests/TestExpectations: Progressions

Canonical link: <a href="https://commits.webkit.org/291726@main">https://commits.webkit.org/291726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8cff159c598eb83057590104050c3997f4d698e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89341 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44180 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94327 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40102 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91392 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9253 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17114 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68832 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26498 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92343 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7102 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81087 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49193 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6848 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36106 "Found 3 new test failures: fonts/monospace.html fonts/sans-serif.html fonts/serif.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39209 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77204 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36457 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96156 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16521 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12141 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77704 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16777 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76878 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77005 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19906 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21427 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20031 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9671 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16535 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21846 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16276 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19727 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18057 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann reviewed (commented):
Looks fine to me -- did you check how LBSE behaves, too?

Ahmad-S792 committed:
Fix SVG paint server fallback handling for a non-existent URL

https://bugs.webkit.org/show_bug.cgi?id=287374
rdar://144493507

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src.git/+/08f435e092b8e4d8ee158c532489aaab08f8d0c4 ,
https://chromium.googlesource.com/chromium/src.git/+/8a0b30e1432898d243c79bab84b373243edb97f2 ,
https://chromium.googlesource.com/chromium/src.git/+/9ba199165cab706b90519a681f67ecce92241d5d

This patch aligns and simplify SVG Paint Server handling as per web specification [1]:

"If a paint server reference in a <paint> is invalid, and no fall-back value is given,
no paint is rendered for that layer."

&

"The value none is a fallback value to use if the paint server reference in the layer
is invalid (due to pointing to an element that does not exist or which is not a
valid paint server)."

[1] https://www.w3.org/TR/SVG2/painting.html#SpecifyingPaint

This also syncs WPT from upstream for "fill-fallback-invalid-uri" test:

Upstream commit: https://github.com/web-platform-tests/wpt/commit/5e030604129f588749fba9d39241edcfe9df1d8c

* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResource.cpp:
(WebCore::requestPaintingResource):
(WebCore::inheritColorFromParentStyleIfNeeded): Deleted.
* LayoutTests/platform/glib/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/glib/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/platform/glib/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/platform/gtk/svg/custom/invalid-fill-expected.png: Removed.
* LayoutTests/platform/gtk/svg/custom/invalid-uri-stroke-expected.png: Removed.
* LayoutTests/platform/gtk/svg/hixie/error/003-expected.png: Removed.
* LayoutTests/platform/ios/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/ios/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/platform/ios/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-fill-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-uri-stroke-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/hixie/error/003-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/invalid-fill-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/custom/invalid-uri-stroke-expected.png: Removed.
* LayoutTests/platform/mac-sonoma-wk2-pixel/svg/hixie/error/003-expected.png: Removed.
* LayoutTests/platform/mac/svg/custom/invalid-fill-expected.png: Removed.
* LayoutTests/platform/mac/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/mac/svg/custom/invalid-uri-stroke-expected.png: Removed.
* LayoutTests/platform/mac/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/platform/mac/svg/hixie/error/003-expected.png: Removed.
* LayoutTests/platform/mac/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/platform/win/svg/custom/invalid-fill-expected.txt: Removed.
* LayoutTests/platform/win/svg/custom/invalid-uri-stroke-expected.txt: Removed.
* LayoutTests/svg/custom/invalid-fill.svg: Removed.
* LayoutTests/svg/custom/invalid-uri-stroke.svg: Removed.
* LayoutTests/svg/custom/recursive-pattern.svg:
* LayoutTests/svg/hixie/error/003-expected.txt: Removed.
* LayoutTests/svg/hixie/error/003.xml: Removed.
* LayoutTests/platform/gtk/svg/custom/external-paintserver-reference-expected.png: Removed.
* LayoutTests/svg/custom/external-paintserver-reference-expected.txt: Rebaselined
* LayoutTests/platform/glib/svg/custom/recursive-pattern-expected.png: Removed.
* LayoutTests/platform/glib/svg/custom/recursive-pattern-expected.txt:
* LayoutTests/platform/gtk/svg/custom/recursive-pattern-expected.png: Removed.
* LayoutTests/platform/ios/svg/custom/recursive-pattern-expected.txt:
* LayoutTests/platform/mac/svg/custom/recursive-pattern-expected.png: Removed.
* LayoutTests/platform/mac/svg/custom/recursive-pattern-expected.txt:
* LayoutTests/TestExpectations: Progressions

Canonical link: https://commits.webkit.org/291726@main
